x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -142,6 +156,113 @@ public void createApiKey(Authentication authentication, CreateApiKeyRequest requ
         }
     }
 
+    /**
+     * Checks for the presence of a {@code Authorization} header with a value that starts with
+     * {@code ApiKey }. If found this will attempt to authenticate the key.
+     */
+    void authenticateWithApiKeyIfPresent(ThreadContext ctx, ActionListener<AuthenticationResult> listener) {
+        if (enabled) {
+            final ApiKeyCredentials credentials;
+            try {
+                credentials = getCredentialsFromHeader(ctx);
+            } catch (ElasticsearchSecurityException ese) {
+                listener.onResponse(AuthenticationResult.terminate(ese.getMessage(), ese));
+                return;
+            }
+
+            if (credentials != null) {
+                final GetRequest getRequest = client.prepareGet(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, credentials.getId())
+                    .setFetchSource(true).request();
+                executeAsyncWithOrigin(ctx, SECURITY_ORIGIN, getRequest, ActionListener.<GetResponse>wrap(response -> {
+                    if (response.isExists()) {
+                        try (ApiKeyCredentials ignore = credentials) {
+                            validateApiKeyCredentials(response.getSource(), credentials, clock, listener);
+                        }
+                    } else {
+                        credentials.close();
+                        listener.onResponse(AuthenticationResult.unsuccessful("unable to authenticate", null));
```

Comment:
Why is a malformed ApiKey (missing colon) a `terminate` but a non-existent id is `unsuccessful`?
I feel like we should either say
- "If there is an ApiKey header, we will not try any other authc method", or
- "If ApiKey fails we fall back to the standard authc chain".

But this looks like a mix of those.

Reply:
Good catch. I went with fall back to the standard chain